core/src/main/java/org/apache/iceberg/data/FormatModelRegistry.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.iceberg.encryption.EncryptedOutputFile;
+import org.apache.iceberg.io.DataWriter;
+import org.apache.iceberg.io.FileAppender;
+import org.apache.iceberg.io.FormatModel;
+import org.apache.iceberg.io.InputFile;
+import org.apache.iceberg.io.ReadBuilder;
+import org.apache.iceberg.io.WriteBuilder;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.util.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A registry that manages file-format-specific readers and writers through a unified object model
+ * factory interface.
+ *
+ * <p>This registry provides access to {@link ReadBuilder}s for data consumption and various writer
+ * builders:
+ *
+ * <ul>
+ *   <li>{@link WriteBuilder} for basic file writing,
+ *   <li>{@link DataWriteBuilder} for data files,
+ *   <li>{@link EqualityDeleteWriteBuilder} for equality deletes,
+ *   <li>{@link PositionDeleteWriteBuilder} for position deletes.
+ * </ul>
+ *
+ * The appropriate builder is selected based on {@link FileFormat} and object model name.
+ *
+ * <p>{@link FormatModel} objects are registered through {@link #registerFormatModel(FormatModel)}
+ * and used for creating readers and writers. Read builders are returned directly from the factory.
+ * Write builders may be wrapped in specialized content file writer implementations depending on the
+ * requested builder type.
+ */
+public final class FormatModelRegistry {
+  private static final Logger LOG = LoggerFactory.getLogger(FormatModelRegistry.class);
+  // The list of classes which are used for registering the reader and writer builders
+  private static final List<String> CLASSES_TO_REGISTER = ImmutableList.of();
+
+  private static final Map<Pair<FileFormat, String>, FormatModel<?>> FORMAT_MODELS =
+      Maps.newConcurrentMap();
+
+  /**
+   * Registers an {@link FormatModel} in this registry.
+   *
+   * <p>The {@link FormatModel} creates readers and writers for a specific combinations of file
+   * format (Parquet, ORC, Avro) and object model (for example: "generic", "spark", "flink", etc.).
+   * Registering custom factories allows integration of new data processing engines for the
+   * supported file formats with Iceberg's file access mechanisms.
+   *
+   * <p>Each factory must be uniquely identified by its combination of file format and object model
+   * name. This uniqueness constraint prevents ambiguity when selecting factories for read and write
+   * operations.
+   *
+   * @param formatModel the factory implementation to register
+   * @throws IllegalArgumentException if a factory is already registered for the combination of
+   *     {@link FormatModel#format()} and {@link FormatModel#modelName()}
+   */
+  @SuppressWarnings("CatchBlockLogException")
+  public static void registerFormatModel(FormatModel<?> formatModel) {
```

Comment:
nit: maybe just `register()` as everything is already contained already in the class name or in the parameter name

Reply:
renamed